metadata-ingestion/src/datahub/ingestion/source/sql/sql_utils.py

```diff
@@ -0,0 +1,296 @@
+    dataset_name: str,
+    domain_config: Dict[str, AllowDenyPattern],
+    domain_registry: DomainRegistry,
+) -> Optional[str]:
+    domain_urn: Optional[str] = None
+
+    domain: str
+    pattern: AllowDenyPattern
+    for domain, pattern in domain_config.items():
+        if pattern.allowed(dataset_name):
+            domain_urn = make_domain_urn(domain_registry.get_domain_urn(domain))
+
+    return domain_urn
+
+
+def gen_database_container(
+    config: SQLAlchemyConfig,
+    database: str,
+    sub_types: List[str],
+    platform: Optional[str] = None,
+    domain_registry: Optional[DomainRegistry] = None,
+    platform_instance: Optional[str] = None,
+    env: Optional[str] = None,
+    report: Optional[SourceReport] = None,
+    database_container_key: Optional[PlatformKey] = None,
+    name: Optional[str] = None,
+    description: Optional[str] = None,
+    owner_urn: Optional[str] = None,
+    external_url: Optional[str] = None,
+    tags: Optional[List[str]] = None,
+    qualified_name: Optional[str] = None,
+    created: Optional[int] = None,
+    last_modified: Optional[int] = None,
+    extra_properties: Optional[Dict[str, str]] = None,
+) -> Iterable[MetadataWorkUnit]:
+    domain_urn: Optional[str] = None
+    if domain_registry:
+        domain_urn = gen_domain_urn(
+            database, domain_config=config.domain, domain_registry=domain_registry
+        )
+
+    if not platform:
+        platform = config.platform
+
+    assert platform
+
+    if not platform_instance:
+        platform_instance = config.platform_instance
+
+    if not env:
+        env = config.env
+
+    database_container_key = (
+        gen_database_key(
+            database, platform=platform, platform_instance=platform_instance, env=env
+        )
+        if not database_container_key
+        else database_container_key
+    )
+
+    container_workunits = gen_containers(
+        container_key=database_container_key,
+        name=name if name else database,
+        sub_types=sub_types,
+        domain_urn=domain_urn,
+        external_url=external_url,
+        description=description,
+        created=created,
+        last_modified=last_modified,
+        tags=tags,
+        owner_urn=owner_urn,
+        qualified_name=qualified_name,
+        extra_properties=extra_properties,
+    )
+
+    for wu in container_workunits:
+        if report:
+            report.report_workunit(wu)
+        yield wu
+
+
+def add_table_to_schema_container(
+    config: SQLAlchemyConfig,
+    dataset_urn: str,
+    db_name: str,
+    schema: str,
+    platform: Optional[str] = None,
+    platform_instance: Optional[str] = None,
+    env: Optional[str] = None,
+    report: Optional[SourceReport] = None,
+    schema_container_key: Optional[PlatformKey] = None,
+) -> Iterable[MetadataWorkUnit]:
+
+    if not platform:
+        platform = config.platform
+
+    assert platform
+
+    if not platform_instance:
+        platform_instance = config.platform_instance
```

Comment:
ideally remove platform, platform_instance, and env options - pull from the key_container instead